### PR TITLE
[ETH-245] TokenGated Rooms: ERC20

### PR DESCRIPTION
**Pull request by bearni95**
- room properties modal now includes info about the token (name, symbol, address) and the minimum required balance
- disabled selector to change room privacy to and from TokenGated
- re-implemented erc20 tokenGates using sagas instead of regular functions

### CODE REVIEW

File: src/components/modals/AddRoomModal.tsx [hunk 1]
```diff
@@ -173,7 +232,48 @@ export default function AddRoomModal({ setOpen, ...props }: ModalProps) {
...
+                                />
+                            </>
+                        )}
+
+                        {tokenType.isCountable && (
+                            <>
+                                <Label htmlFor="minTokenAmount">Minimum Token Amount</Label>
+                                <TextField
+                                    id="minTokenAmount"
+                                    value={minTokenAmount}
```
Comment (mondoreale) [inline src/components/modals/AddRoomModal.tsx:268]: `minTokenAmount` holds a number and `parseFloat('')` is a `NaN`. If you remove the initial 0 the field will display an unchangeable `NaN`. Idea:
- let's hold and use a string here, and
- hold its numeric value in a separate consts, and
- use that numeric value in your logic (zero when invalid).

File: src/components/modals/RoomPropertiesModal.tsx [hunk 2]
```diff
@@ -95,8 +94,40 @@ export default function RoomPropertiesModal({ open, setOpen, ...props }: ModalPr
...
+
+    useEffect(() => {
+        if (!tokenAddress || !tokenType) {
+            return
+        }
+
+        const metadata = getTokenMetadata()
+        console.log({ metadata })
+        console.log({ tokenAddress, tokenId, minTokenAmount, tokenType })
+    }, [tokenAddress, tokenId, minTokenAmount, tokenType])*/
```
Comment (mondoreale) [inline src/components/modals/RoomPropertiesModal.tsx:130]: Can we remove it?

File: src/features/member/sagas/promoteDelegatedAccount.saga.ts [hunk 3]
```diff
@@ -1,14 +1,42 @@
...
+            console.log('promoting token gated room')
+            const stream: EnhancedStream = yield streamrClient.getStream(roomId)
+            const metadata = stream.extensions['thechat.eth']
+
+            if (!metadata.tokenAddress) {
+                throw new Error('No token address found')
+            }
+            console.log({ metadata })
+
+            const owner: Address = yield streamrClient.getAddress()
```
Comment (mondoreale) [inline src/features/member/sagas/promoteDelegatedAccount.saga.ts:26]: Value of `streamrClient.getAddress()` should be equal to the value passed as `requester`. Maybe `requester` is redundant… I like the `getAddress` way. Sounds like a future "todo".

File: src/features/room/hooks.ts [hunk 4]
```diff
@@ -37,7 +41,18 @@ export function usePrivacy(roomId: undefined | RoomId) {
...
+    switch (privacy) {
+        case PrivacySetting.Public:
+            return PublicRoomOption
+        case PrivacySetting.Private:
+            return PrivateRoomOption
+        case PrivacySetting.TokenGated:
+            return TokenGatedRoomOption
+        default:
+            return PrivateRoomOption
+    }
```
Comment (mondoreale) [inline src/features/room/hooks.ts:55]:
```suggestion
    switch (privacy) {
        case PrivacySetting.Public:
            return PublicRoomOption
        case PrivacySetting.TokenGated:
            return TokenGatedRoomOption
        case PrivacySetting.Private:
        default:
            return PrivateRoomOption
    }
```

File: src/features/room/sagas/create.saga.ts [hunk 5]
```diff
@@ -55,6 +57,24 @@ function* onCreateAction({
             },
         } as StreamProperties)
 
+        if (
+            privacy === PrivacySetting.TokenGated &&
+            metadata.tokenType!.standard === TokenTypes.ERC20.standard &&
+            metadata.tokenAddress &&
+            metadata.minTokenAmount
```
Comment (mondoreale) [inline src/features/room/sagas/create.saga.ts:64]: Do we accept negative `minTokenAmount` values? Can a negative value be entered?

File: src/features/room/sagas/getPrivacy.saga.ts [hunk 6]
```diff
@@ -1,31 +1,40 @@
...
 }: ReturnType<typeof RoomAction.getPrivacy>) {
     try {
-        const stream: undefined | Stream = yield getStream(streamrClient, roomId)
+        const stream: undefined | EnhancedStream = yield getStream(streamrClient, roomId)
 
         if (!stream) {
             throw new RoomNotFoundError(roomId)
         }
 
+        const isTokenGated: boolean = yield isTokenGatedRoom(roomId, streamrClient)
```
Comment (mondoreale) [inline src/features/room/sagas/getPrivacy.saga.ts:21]: Can we have an alternative (synchronous maybe) implementation of this check to skip the additional fetching? We already hold a copy of \*the\* `stream`.

File: src/features/member/sagas/promoteDelegatedAccount.saga.ts [hunk 7]
```diff
@@ -1,14 +1,42 @@
...
+import { put } from 'redux-saga/effects'
 import { StreamPermission } from 'streamr-client'
 
 function* onPromoteDelegatedAccountAction({
-    payload: { roomId, delegatedAddress, provider, requester, streamrClient },
+    payload: { roomId, delegatedAddress, provider, requester, streamrClient, privacy },
 }: ReturnType<typeof MemberAction.promoteDelegatedAccount>) {
     try {
+        // split for token-gated rooms
+        if (privacy === PrivacySetting.TokenGated) {
```
Comment (mondoreale) [inline src/features/member/sagas/promoteDelegatedAccount.saga.ts:16]: Should we go with 2 different actions, 1 for token-gated rooms and another one for regular rooms? Feels like it'd add clarity.

PS. Naming is hard. It'd have to be something like `onTokenGatedPromoteDelegatedAccountAction`. 🤔

File: src/features/room/sagas/pin.saga.ts [hunk 8]
```diff
@@ -26,6 +35,19 @@ function* pinRemote(owner: Address, roomId: RoomId, { streamrClient }: PinRemote
 
     const metadata = stream.extensions['thechat.eth']
 
+    const isTokenGated: boolean = yield isTokenGatedRoom(roomId, streamrClient)
+
+    if (isTokenGated && metadata.tokenAddress) {
```
Comment (mondoreale) [inline src/features/room/sagas/pin.saga.ts:40]: What's the logic here? What does a "token gated" room mean in a pin situation? I join if I meet the conditions?

File: src/features/tokenGatedRooms/index.ts [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+import { createAction, createReducer } from '@reduxjs/toolkit'
+import { all } from 'redux-saga/effects'
+import StreamrClient, { BigNumber } from 'streamr-client'
+import registerERC20Policy from '$/features/tokenGatedRooms/sagas/registerERC20Policy.saga'
+import joinERC20 from '$/features/tokenGatedRooms/sagas/joinERC20.saga'
+import getTokenMetadata from '$/features/tokenGatedRooms/sagas/getTokenMetadata.saga'
+import { RoomId } from '$/features/room/types'
+
+const initialState: TokenGatedRoomState = {
+    selectedRoomId: undefined,
```
Comment (mondoreale) [inline src/features/tokenGatedRooms/index.ts:14]: Utilized anywhere?

File: src/features/tokenGatedRooms/sagas/getTokenMetadata.saga.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+        yield put(
+            TokenGatedRoomAction.setERC20Metadata({
+                name,
+                symbol,
+                decimals,
+            })
+        )
+    } catch (e) {
+        console.error(e)
```
Comment (mondoreale) [inline src/features/tokenGatedRooms/sagas/getTokenMetadata.saga.ts:31]: Let's use `handleError(e)`.

File: src/features/tokenGatedRooms/sagas/joinERC20.saga.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+        const factory = getJoinPolicyFactory(provider)
+        const policyAddress: Address = yield factory.erc20TokensToJoinPolicies(tokenAddress, roomId)
+        const policy = getERC20JoinPolicy(policyAddress, provider)
+        const canJoin: boolean = yield policy.canJoin(owner)
+        if (!canJoin) {
+            throw new Error('Cannot join tokenGated room')
+        }
+        yield policy.requestDelegatedJoin(delegatedAccount)
+    } catch (e) {
+        console.error(e)
```
Comment (mondoreale) [inline src/features/tokenGatedRooms/sagas/joinERC20.saga.ts:20]: `handleError`

File: src/features/tokenGatedRooms/sagas/registerERC20Policy.saga.ts [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        // assign permissions to the freshly-deployed policy contract
+        yield stream.grantPermissions({
+            user: policyAddress,
+            permissions: [StreamPermission.GRANT],
+        })
+
+        // give up one's own control over the stream
+        // for some reason forces the owner to pin the room as if they were an invitee
+        toastId = createInformationalToast('Revoking GRANT permission for the owner', toastId)
+        yield stream.revokePermissions({
```
Comment (mondoreale) [inline src/features/tokenGatedRooms/sagas/registerERC20Policy.saga.ts:78]: Can we do the granting and the revoking in a single call using `setMultiplePermissions`?

File: src/features/tokenGatedRooms/utils/isTokenGatedRoom.ts [hunk 13]
```diff
@@ -0,0 +1,18 @@
+import { EnhancedStream } from '$/types'
+import getStream from '$/utils/getStream'
+import RoomNotFoundError from '$/errors/RoomNotFoundError'
+import { RoomId } from '$/features/room/types'
+import { StreamrClient } from 'streamr-client'
+
+export async function isTokenGatedRoom(roomId: RoomId, streamrClient: StreamrClient) {
```
Comment (mondoreale) [inline src/features/tokenGatedRooms/utils/isTokenGatedRoom.ts:7]: We use it in `pin.saga` and `getPrivacy` saga. In both cases we already have `roomId`-represented stream at hand.

Let's turn it into
```ts
export default function isTokenGatedRoom(stream: EnhancedStream) {
    return !!stream.extensions['thechat.eth'].tokenAddress
}
```

**Review by mondoreale — COMMENTED** [review]
PR looks very promising. 👍🏻 Left a couple of comments.

File: src/components/modals/AddRoomModal.tsx [hunk 14]
```diff
@@ -41,58 +45,99 @@ export default function AddRoomModal({ setOpen, ...props }: ModalProps) {
...
+        setTokenId(0)
+        setMinTokenAmount(0)
     }
 
     const provider = useWalletProvider()
 
     const streamrClient = useWalletClient()
 
-    function onSubmitCreate() {
+    async function onSubmitCreate() {
```
[the remark this answers is not in the document]
Comment (mondoreale) [reply src/components/modals/AddRoomModal.tsx:79]: This isn't very important and is will most likely change when the actual design gets implemented. Ignore it.

File: src/features/member/sagas/promoteDelegatedAccount.saga.ts (hunk 7 above)
Comment (bearni95) [reply src/features/member/sagas/promoteDelegatedAccount.saga.ts:18]: Sounds like a sensitive option, i've left the split inside the `usePromoteDelegatedAccount` hook

File: src/features/room/sagas/pin.saga.ts (hunk 8 above)
Comment (bearni95) [reply src/features/room/sagas/pin.saga.ts]: Exactly! From the perspective of a user who wants to join a token-gated room, pinning felt like the most natural action to attach it to.

File: src/features/tokenGatedRooms/index.ts (hunk 9 above)
Comment (bearni95) [reply src/features/tokenGatedRooms/index.ts:14]: Only to create the reducer

File: src/components/modals/AddRoomModal.tsx [hunk 15]
```diff
@@ -41,58 +45,92 @@ export default function AddRoomModal({ setOpen, ...props }: ModalProps) {
...
+                        createdBy: account!,
+                        id: `/${Prefix.Room}/${uuidv4()}`,
+                        name: roomName,
+                        owner: account!,
+                        updatedAt: now,
+                        tokenAddress,
+                        tokenId,
+                        minTokenAmount: isNaN(parseFloat(minTokenAmount))
+                            ? 0
+                            : parseFloat(minTokenAmount),
```
Comment (mondoreale) [inline src/components/modals/AddRoomModal.tsx:102]: How about `minTokenAmount: parseFloat(minTokenAmount) || 0`?

**Review by mondoreale — APPROVED** [review]
Looks good. 👍🏻 

We talked briefly about `TestERC*` files; that we don't need them. Let's clean it up in case that statement holds up.

I'll test the feature more extensively while I'm implementing the new UI. :)

File: src/features/tokenGatedRooms/index.ts (hunk 9 above)
Comment (mondoreale) [reply src/features/tokenGatedRooms/index.ts:14]: Is `selectedRoomId` used anywhere? ;)

**Pull request merged**